File: libraries/kafka-client-py/kafka_client_py/consumer/dedupe/gate.py
import logging
from collections.abc import Callable

from ..message import ConsumedMessage
from .store import DedupeStore

logger = logging.getLogger(__name__)


EventIdExtractor = Callable[[ConsumedMessage], str | None]


class DedupeGate:
    def __init__(
        self,
        dedupe_store: DedupeStore | None,
        event_id_extractor: EventIdExtractor | None,
    ) -> None:
        self._dedupe_store = dedupe_store
        self._event_id_extractor = event_id_extractor

    async def already_processed(self, message: ConsumedMessage) -> bool:
        if self._dedupe_store is None or self._event_id_extractor is None:
            return False

        event_id = self._event_id_extractor(message)
        if event_id is None:
            return False

        if await self._dedupe_store.seen(event_id):
            logger.debug("kafka.dedupe.skip", extra={"event_id": event_id})
            return True
        return False

File: libraries/kafka-client-go/consumer/dedupe/gate.go
package dedupe

import (
	"context"
	"log/slog"

	kafkaclient "github.com/power-finance/kafka-client-go"
)

type EventIDExtractor func(message kafkaclient.ConsumedMessage) (eventID string, found bool)

type Gate struct {
	store     Store
	extractor EventIDExtractor
	logger    *slog.Logger
}

func NewGate(store Store, extractor EventIDExtractor, logger *slog.Logger) *Gate {
	if logger == nil {
		logger = slog.Default()
	}

	storeConfigured := store != nil
	extractorConfigured := extractor != nil
	if storeConfigured != extractorConfigured {
		logger.Warn(
			"kafka.dedupe.disabled_partial_config",
			slog.Bool("store_configured", storeConfigured),
			slog.Bool("extractor_configured", extractorConfigured),
		)
	}

	return &Gate{
		store:     store,
		extractor: extractor,
		logger:    logger,
	}
}

func (g *Gate) AlreadyProcessed(
	ctx context.Context,
	message kafkaclient.ConsumedMessage,
) (bool, error) {
	if g.store == nil || g.extractor == nil {
		return false, nil
	}

	eventID, isFound := g.extractor(message)
	if !isFound || eventID == "" {
		return false, nil
	}

	isSeen, seenErr := g.store.Seen(ctx, eventID)
	if seenErr != nil {
		return false, seenErr
	}

	if isSeen {
		g.logger.DebugContext(ctx, "kafka.dedupe.skip", slog.String("event_id", eventID))
		return true, nil
	} else {
		return false, nil
	}
}

// MarkProcessed records the message's event id as consumed so a redelivery is
// skipped by AlreadyProcessed; a no-op without a store/extractor or event id.
func (g *Gate) MarkProcessed(
	ctx context.Context,
	message kafkaclient.ConsumedMessage,
) error {
	if g.store == nil || g.extractor == nil {
		return nil
	}

	eventID, isFound := g.extractor(message)
	if !isFound || eventID == "" {
		return nil
	}

	return g.store.Mark(ctx, eventID)
}
